### PR TITLE
Feature allows to send diff-File (as File-object) to spyware-server. 

### PR DESCRIPTION
**Pull request by samutamm**
@wakr @samutamm

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -10,31 +10,22 @@ tmc-cli.iml
 /scripts/config
 /scripts/*.jar
 config.properties
-<<<<<<< HEAD
```
Comment (samutamm) [inline .gitignore]: remove HEAD-tag

File: testResources/successExercise/viikko1/Viikko1_001.Nimi/nbproject/private/private.xml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (samutamm) [inline testResources/successExercise/viikko1/Viikko1_001.Nimi/nbproject/private/private.xml]: ignore this file

File: testResources/successExercise/viikko1/Viikko1_001.Nimi/build_log.txt [hunk 3]
```diff
@@ -42,12 +42,12 @@
 init:
 
 -deps-jar-init:
```
Comment (samutamm) [inline testResources/successExercise/viikko1/Viikko1_001.Nimi/build_log.txt]: ignore?

File: testResources/successExercise/viikko1/Viikko1_001.Nimi/build/built-jar.properties [hunk 4]
```diff
@@ -1,4 +0,0 @@
-#Mon, 01 Jun 2015 15:06:27 +0300
```
Comment (samutamm) [inline testResources/successExercise/viikko1/Viikko1_001.Nimi/build/built-jar.properties]: ignore?

File: testResources/halfdoneExercise/viikko1/Viikko1_004.Muuttujat/build_log.txt [hunk 5]
```diff
@@ -42,12 +42,12 @@
 init:
```
Comment (samutamm) [inline testResources/halfdoneExercise/viikko1/Viikko1_004.Muuttujat/build_log.txt]: ignore?

File: src/main/java/hy/tmc/cli/spyware/DiffSender.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+
+public class DiffSender {
+
+    /**
+     * Sends given difffile to all URLs specified by course.
```
Comment (samutamm) [inline src/main/java/hy/tmc/cli/spyware/DiffSender.java]: Modify javadoc and add param tags.

File: src/main/java/hy/tmc/cli/backend/communication/UrlCommunicator.java [hunk 7]
```diff
@@ -126,12 +133,25 @@ private static void addCredentials(HttpRequestBase httpRequest, String credentia
...
     }
 
-    private static HttpResult getResponseResult(HttpRequestBase httpRequest) 
+    /**
+     * Adds headers to request if present.
+     *
+     * @param httpRequest where to put headers. 
+     * @param headers to be included.
+     */
+    private static void addHeadersTo(HttpRequestBase httpRequest, Map<String, String> headers) {
```
Comment (jamox) [inline src/main/java/hy/tmc/cli/backend/communication/UrlCommunicator.java]: Maybe avoid passing null, and use Optional instead

File: src/main/java/hy/tmc/cli/backend/communication/CourseSubmitter.java [hunk 8]
```diff
@@ -85,28 +85,27 @@ public boolean isExpired(Exercise currentExercise){
...
+        return date.getTime() > current.getTime();
     }
 
     private String sendSubmissionToServerWithPaste(
             String submissionZipPath,
             String url) throws IOException {
         HttpResult result = UrlCommunicator.makePostWithFile(
-                new File(submissionZipPath), url + "&paste=1"
+                new File(submissionZipPath),
+                url + "&paste=1",
```
Comment (wakr) [inline src/main/java/hy/tmc/cli/backend/communication/CourseSubmitter.java]: magic

File: src/main/java/hy/tmc/cli/backend/communication/CourseSubmitter.java [hunk 9]
```diff
@@ -85,28 +85,27 @@ public boolean isExpired(Exercise currentExercise){
...
+                Optional.<Map<String, String>>absent()
         );
         return TmcJsonParser.getPasteUrl(result);
     }
 
     private String sendZipFile(String currentPath, Exercise currentExercise, boolean paste) throws IOException {
-        String submissionZipPath = currentPath + "/submission.zip";
+        this.submissionZipPath = currentPath + "/submission.zip";
         String returnUrl = currentExercise.getReturnUrlWithApiVersion();
```
Comment (wakr) [inline src/main/java/hy/tmc/cli/backend/communication/CourseSubmitter.java]: Refactor this

File: src/main/java/hy/tmc/cli/backend/communication/CourseSubmitter.java [hunk 10]
```diff
@@ -139,8 +137,9 @@ private Exercise findExercise(String currentPath) {
     private List<Exercise> findCourseExercises(String currentPath) {
         Course currentCourse = getCurrentCourse(currentPath);
         if (currentCourse == null) {
```
Comment (wakr) [inline src/main/java/hy/tmc/cli/backend/communication/CourseSubmitter.java]: Refactor to use optional

File: src/main/java/hy/tmc/cli/backend/communication/CourseSubmitter.java [hunk 11]
```diff
@@ -203,4 +202,18 @@ public Course findCourseByPath(String[] foldersPath) {
...
         return directoryPath.split("/");
     }
+
+    /**
+     * If class submissionZipPath is defined in sendZipFile-method 
+     * and the file in defined path exists, it will be removed.
+     * This method should be invoked allways when submit-function fails.
+     */
+    private void deleteZipIfExists() {
+        if (submissionZipPath != null) {
```
Comment (wakr) [inline src/main/java/hy/tmc/cli/backend/communication/CourseSubmitter.java]: https://code.google.com/p/guava-libraries/wiki/UsingAndAvoidingNullExplained

File: src/main/java/hy/tmc/cli/backend/communication/UrlCommunicator.java [hunk 12]
```diff
@@ -126,12 +133,26 @@ private static void addCredentials(HttpRequestBase httpRequest, String credentia
...
 
-    private static HttpResult getResponseResult(HttpRequestBase httpRequest) 
+    /**
+     * Adds headers to request if present.
+     *
+     * @param httpRequest where to put headers.
+     * @param headers to be included.
+     */
+    private static void addHeadersTo(HttpRequestBase httpRequest, Optional<Map<String, String>> headers) {
+        System.out.println("HEADERS: chang: " + headers);
```
Comment (wakr) [inline src/main/java/hy/tmc/cli/backend/communication/UrlCommunicator.java]: Remove sout

File: testResources/2013_ohpeJaOhja/viikko1/Viikko1_002.HeiMaailma/nbproject/private/private.xml [hunk 13]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (wakr) [inline testResources/2013_ohpeJaOhja/viikko1/Viikko1_002.HeiMaailma/nbproject/private/private.xml]: IGNORE

File: testResources/failingExercise/viikko1/Viikko1_001.Nimi/build/built-jar.properties [hunk 14]
```diff
@@ -0,0 +1,4 @@
+#Wed, 03 Jun 2015 15:35:37 +0300
```
Comment (wakr) [inline testResources/failingExercise/viikko1/Viikko1_001.Nimi/build/built-jar.properties]: IGNORE

File: testResources/halfdoneExercise/viikko1/Viikko1_004.Muuttujat/build/built-jar.properties [hunk 15]
```diff
@@ -1,4 +1,4 @@
-#Wed, 03 Jun 2015 14:35:56 +0300
```
Comment (wakr) [inline testResources/halfdoneExercise/viikko1/Viikko1_004.Muuttujat/build/built-jar.properties]: IGNORE

File: testResources/halfdoneExercise/viikko1/Viikko1_004.Muuttujat/build_log.txt [hunk 16]
```diff
@@ -0,0 +1,85 @@
+
```
Comment (wakr) [inline testResources/halfdoneExercise/viikko1/Viikko1_004.Muuttujat/build_log.txt]: IGNORE

File: testResources/successExercise/viikko1/Viikko1_001.Nimi/build/built-jar.properties [hunk 17]
```diff
@@ -0,0 +1,4 @@
+#Wed, 03 Jun 2015 15:35:37 +0300
```
Comment (wakr) [inline testResources/successExercise/viikko1/Viikko1_001.Nimi/build/built-jar.properties]: IGNORE

File: testResources/successExercise/viikko1/Viikko1_001.Nimi/build_log.txt [hunk 18]
```diff
@@ -0,0 +1,85 @@
+
+-pre-init:
```
Comment (wakr) [inline testResources/successExercise/viikko1/Viikko1_001.Nimi/build_log.txt]: IGNORE

**Pull request merged**